bam_fastq.c

```diff
@@ -880,135 +578,296 @@ static inline bool filter_it_out(const bam1_t *b, const bam2fq_state_t *state)
-                    } else {
-                        if (score[1] > 0) {
-                            if (bgzf_write(state->fpr[1], linebuf[1].s, linebuf[1].l) < 0) { valid = false; break; }
-                        } else {
-                            if (bgzf_write(state->fpr[2], linebuf[2].s, linebuf[2].l) < 0) { valid = false; break; }
-                        }
-                    }
-                }
-                if (score[0]) { // TODO: check this
-                    // print linebuf[0] to fpr[0]
-                    if (bgzf_write(state->fpr[0], linebuf[0].s, linebuf[0].l) < 0) { valid = false; break; }
-                }
+        if (!at_eof) {
+            ++n_reads;
+
+            // Handle -O option: use OQ for qual
+            uint8_t *oq;
+            if (state->use_oq && (oq = bam_aux_get(b[n],"OQ")) && *oq == 'Z') {
+                int i, l = strlen((char *)++oq);
+                uint8_t *qual = bam_get_qual(b[n]);
+                for (i = 0; i < l && i < b[n]->core.l_qseq; i++)
+                    qual[i] = oq[i] - '!';
             }
+        }
 
+        if (at_eof
+            || !current_qname
+            || (strcmp(current_qname, bam_get_qname(b[n])) != 0)) {
+            // New name, so flush best examples of previous name.
+            if (current_qname)
+                if (flush_rec(state, opts, b, score, best, &n_singletons) < 0)
+                    goto err;
 
-            free(current_qname); current_qname = NULL;
+            current_qname = bam_get_qname(b[n]);
             score[0] = score[1] = score[2] = 0;
-            for (n=0; n < 3; n++) {
-                bam_destroy1(records[n]); records[n]=NULL;
-            }
 
             if (at_eof) { break; }
-
-            current_qname = strdup(bam_get_qname(b));
-            if (!current_qname) { valid = false; break; }
         }
 
         // Prefer a copy of the read that has base qualities
-        int b_score = bam_get_qual(b)[0] != 0xff? 2 : 1;
-        readpart rp = which_readpart(b);
-        if (b_score > score[rp]) {
-            if (!tags2fq(b, state, opts)) { valid = false; break; }
-            if (records[rp]) bam_destroy1(records[rp]);
-            records[rp] = b;
+        int b_score = bam_get_qual(b[n])[0] != 0xff? 2 : 1;
+        readpart rp = which_readpart(b[n]);
+        if (score[rp] < b_score) {
             score[rp] = b_score;
-            b = NULL;
-            if(!bam1_to_fq(records[rp], &linebuf[rp], state)) {
-                fprintf(stderr, "[%s] Error converting read to FASTA/Q\n", __func__);
-                valid = false; break;
-            }
+            // Record b[n] slot for best copy of readpair and find a new
+            // slot for next bam read
+            best[rp] = n;
+            int used_slot[4] = {0}, i;
+            for (i = 0; i < 3; i++)
+                if (best[i] >= 0)
+                    used_slot[best[i]] = 1;
+            for (i = 0; i < 6 && used_slot[i]; i++)
```

Comment:
The 6 here was missed when changing the size of `used_slot`.

Reply:
Oops.  Frankly surprised the compiler didn't complain at that one!